add $syntax_check() and tests

rok-cesnovar opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Fixes #276 

@paul-buerkner if you have a few minutes, you can test with

`remotes::install_github("stan-dev/cmdstanr@syntax_check")
`
```r
library(cmdstanr)

file_path <- file.path(cmdstan_path(), "examples", "bernoulli", "bernoulli.stan")
mod <- cmdstan_model(file_path, compile = FALSE)
mod$syntax_check()
```

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Univ. of Ljubljana


By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

paul-buerkner commented:
Thanks a lot! Will check it out tonight when I am back on my laptop.

Codecov Comments Bot <notifications@github.com> schrieb am Fr., 14. Aug.
2020, 17:05:

> Codecov <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=h1>
> Report
>
> Merging #277
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=desc> into
> master
> <https://codecov.io/gh/stan-dev/cmdstanr/commit/762375eaab99590484c6965f629127b04fe08d69&el=desc>
> will *decrease* coverage by 0.24%.
> The diff coverage is 76.19%.
>
> [image: Impacted file tree graph]
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=tree>
>
> @@            Coverage Diff             @@
> ##           master     #277      +/-   ##
>
> ==========================================
> - Coverage   90.30%   90.05%   -0.25%
>
> ==========================================
>
>   Files          12       12
>
>   Lines        2372     2414      +42
>
> ==========================================
> + Hits         2142     2174      +32
> - Misses        230      240      +10
>
> Impacted Files
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=tree> Coverage
> Δ
> R/utils.R
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277/diff?src=pr&el=tree#diff-Ui91dGlscy5S> 70.78%
> <66.66%> (-0.08%) ⬇️
> R/model.R
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277/diff?src=pr&el=tree#diff-Ui9tb2RlbC5S> 87.82%
> <76.92%> (-1.56%) ⬇️
> ------------------------------
>
> Continue to review full report at Codecov
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=continue>.
>
> *Legend* - Click here to learn more
> <https://docs.codecov.io/docs/codecov-delta>
> Δ = absolute <relative> (impact), ø = not affected, ? = missing data
> Powered by Codecov
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=footer>. Last
> update 762375e...6d3d9a5
> <https://codecov.io/gh/stan-dev/cmdstanr/pull/277?src=pr&el=lastupdated>.
> Read the comment docs <https://docs.codecov.io/docs/pull-request-comments>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/cmdstanr/pull/277#issuecomment-674091475>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADCW2AEHY5QQDZ3453AYC3LSAVACJANCNFSM4P7QUF5A>
> .
>

jgabry commented:
Awesome. I'll try to review this later today or this weekend.

jgabry commented on R/model.R:
```diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+#' @aliases syntax_check
+#' @family CmdStanModel methods
+#'
+#' @description The `$syntax_check()` method of a [`CmdStanModel`] object
+#'   checks the Stan syntax of the model and returns `TRUE` if the compilation
+#'   succeeds. In case of a syntax error, an error is thrown.
```
```suggestion
#' @description The `$syntax_check()` method of a [`CmdStanModel`] object
#'   checks the Stan program for syntax errors and returns `TRUE` (invisibly) if 
#'   parsing succeeds. In case of a syntax error, an error is thrown.
```

jgabry commented on R/model.R:
````diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+#' @section Usage:
+#'   ```
+#'   $syntax_check(
+#'     quiet = TRUE,
````
Unlike some of the other methods with a `quiet` argument, I'm not sure we need it here. Is there a case when it's helpful to see the command here? If not, I'd suggest removing the argument. If there is, then it's fine to keep it.

rok-cesnovar replied:
quiet = FALSE now outputs "Stan program is syntactically correct"
quiet = TRUE only outputs anything on stderr from the compiler

jgabry commented on R/model.R:
````diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+#'   ```
+#'
+#' @section Arguments:
+#'   * `quiet`: (logical) Should the compile command be printed.
````
See comment above about `quiet`

jgabry commented on R/model.R:
```diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+#'   * `include_paths`: (character vector) Paths to directories where Stan
+#'   should look for files specified in `#include` directives in the Stan
+#'   program.
+#'   * `stanc_options`: (list) Any Stan-to-C++ transpiler options to be used
+#'   when compiling the model. See the **Examples** section below as well as the
+#'   `stanc` chapter of the CmdStan Guide for more details on available options:
+#'   https://mc-stan.org/docs/cmdstan-guide/stanc.html.
```
The **Examples** section referred to here is actually in the doc for the compile method. Maybe change to: 

```suggestion
#'   * `stanc_options`: (list) Any Stan-to-C++ transpiler options to be used
#'   when compiling the model. See the documentation for the
#'   [`$compile()`][model-method-compile] method for details.
```

jgabry commented on R/model.R:
```diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+#'   `stanc` chapter of the CmdStan Guide for more details on available options:
+#'   https://mc-stan.org/docs/cmdstan-guide/stanc.html.
+#'
+#' @section Value: The `$syntax_check()` method returns `TRUE` if the model
```
```suggestion
#' @section Value: The `$syntax_check()` method returns `TRUE` (invisibly) if the model
```

jgabry commented on R/model.R:
```diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
+    stanc_options[["name"]] <- model_name
+  }
+
+  stanc_built_options = c()
+  for (i in seq_len(length(stanc_options))) {
+    option_name <- names(stanc_options)[i]
+    if (isTRUE(as.logical(stanc_options[[i]]))) {
+      stanc_built_options <- c(stanc_built_options, paste0("--", option_name))
+    } else {
+      stanc_built_options <- c(stanc_built_options, paste0("--", option_name, "=", stanc_options[[i]]))
+    }
+  }
```
If possible can you add a test just making sure that this works or at least doesn't error?

rok-cesnovar replied:
stanc_options are used always for at least the name, so this works and is already being tested. Will add a test where we specify an additional flag.

jgabry commented on R/model.R:
```diff
@@ -506,6 +506,114 @@ compile_method <- function(quiet = TRUE,
 
+#' Check syntax of a Stan program
+#'
+#' @name model-method-syntax-check
```
```suggestion
#' @name model-method-syntax_check
```

paul-buerkner commented:
This feature works well! Just one minor question: It seems that the initial message "Syntax checking the Stan program..." is passed via the same stream as the details of the error message if an error occurs. As a result, I cannot turn of the initial message separately. Is the message really required?

rok-cesnovar commented:
> Is this message really required?

Yeah, maybe not. The syntax check is instant (if a model had a few thousand lines it would maybe take a second.

Would this work: text that the syntax is OK on stdout and a compile error printed as message()

paul-buerkner commented:
I think if OK was printed as message, that would also be fine and more consistent, but given the instant execution time, I thought the initial message may not be required.

paul-buerkner commented:
Alternatively, perhaps there could be an argument to silence these messages? For use in brms, it would be the best to say nothing when it worked and throw in the information when it did. I don't want to make these too much about brms though so you go with the approach you think is best for cmdstanr in general.

rok-cesnovar commented:
I think we can reuse quiet for this case. 

quiet = TRUE-> Only show the compile error, otherwise just return TRUE 
quiet  = FALSE -> print any other messages

Will wait for @jgabry to comment on that before I proceed.

> I don't want to make these too much about brms

I think the points that you make about the use in brms will apply for other packages that will use cmdstanr. These are all very good points I think.

jgabry commented:
> quiet = TRUE-> Only show the compile error, otherwise just return TRUE
> quiet = FALSE -> print any other messages

That sounds good to me!

rok-cesnovar commented:
This is actually a nice and fast way of checking a model with the pedantic check without having to compile to the C++ level. So thanks @paul-buerkner for the idea! 

We should use this in a vignette to show how users can run the pedantic check in cmdstanr.

paul-buerkner commented:
Glad it turns out to be useful outside of brms as well!

rok-cesnovar commented:
@jgabry I think I addressed everything. I added a test for the pedantic mode so we test additional flags. Also added the pedantic mode example to the vignette. Not sure if I need to rebuild the vignette or anything. Anyhow, ready for another look.
